internal/reconcilers/cluster.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	infrastructurev1alpha3 "github.com/tinkerbell/cluster-api-provider-tinkerbell/api/v1alpha3"
+)
+
+const (
+	kubernetesAPIPort = 6443
+)
+
+// ReconcileContext describes functionality required for reconciling Machine or Cluster object
+// into Tinkerbell Kubernetes node.
+type ReconcileContext interface {
+	Reconcile() error
+}
+
+// ClusterReconcileContextConfig is a configuration struct for creating clusterReconcileContext.
+type ClusterReconcileContextConfig struct {
+	Client           client.Client
+	HardwareIPGetter hardwareIPGetter
+	Log              logr.Logger
+}
+
+// hardwareIPGetter is a slice of Tinkerbell client functionality we use when reconciling Cluster
+// object.
+type hardwareIPGetter interface {
+	NextAvailableHardwareID(ctx context.Context) (string, error)
+	GetHardwareIP(ctx context.Context, uuid string) (string, error)
+}
+
+// clusterReconcileContext implements ReconcileContext by reconciling TinkerbellCluster object.
+type clusterReconcileContext struct {
+	ctx               context.Context
+	tinkerbellCluster *infrastructurev1alpha3.TinkerbellCluster
+	patchHelper       *patch.Helper
+	hardwareIPGetter  hardwareIPGetter
+	cluster           *clusterv1alpha3.Cluster
+	log               logr.Logger
+	client            client.Client
+}
+
+// New builds a context for cluster reconciliation process, collecting all required
+// information.
+//
+// If unexpected case occurs, error is returned.
+//
+// If some data is not yet available, nil is returned.
+func (crcc *ClusterReconcileContextConfig) New(namespacedName types.NamespacedName) (ReconcileContext, error) {
+	crc := &clusterReconcileContext{
+		log:               crcc.Log.WithValues("tinkerbellcluster", namespacedName),
+		ctx:               context.Background(),
```

Comment:
When we look to updating to support the in-development v1alpha4 version of Cluster API, this is going to get a bit messy with the way controller-runtime v0.7.0 is using contexts and passing loggers around: https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.7.0, the biggest one being that `Reconcile()` will be passed a context and that context will be enriched with a logger.

Reply:
I see. I think when the time comes to an update, we can then:
- Remove the `Log` field from the configuration.
- Accept context as a parameter to `New()`. This can actually be done right now.

Reply:
I added a context parameter to `New()`.